File: ui/Sources/TBKit/Queries.swift
import Foundation

public enum LookupResult: Sendable, Hashable {
    case account(Account)
    case transfer(Transfer)
    case none
}

/// Result of looking up a transfer id from an account's point of view.
public enum AccountTransferMatch: Sendable, Hashable {
    /// The transfer debits or credits the account.
    case onAccount(Transfer)
    /// The transfer exists but moves funds between other accounts.
    case otherAccounts(Transfer)
    case notFound
}

public struct ClusterInfo: Sendable, Hashable {
    public var clusterID: UInt128
    public var addresses: [String]
    public var clientVersion: String
    public var latency: Duration
}

extension TBClient {
    /// Opens a client and verifies the cluster accepts it, bounding the handshake.
    public static func connect(
        clusterID: UInt128, addresses: [String], timeout: Duration = .seconds(5)
    ) async throws(TBError) -> (TBClient, ClusterInfo) {
        let client = try TBClient(clusterID: clusterID, addresses: addresses)
        do {
            let latency = try await client.ping(timeout: timeout)
            let info = ClusterInfo(
                clusterID: clusterID, addresses: addresses,
                clientVersion: clientVersion, latency: latency)
            return (client, info)
        } catch {
            client.close()
            throw error
        }
    }

    /// The first request after init includes session registration, so it is sent as a
    /// warm-up and the second round trip is timed.
    public func ping(timeout: Duration = .seconds(5)) async throws(TBError) -> Duration {
        let probe = Wire.ids([UInt128.max])
        _ = try await submit(.lookupAccounts, payload: probe, timeout: timeout)
        let clock = ContinuousClock()
        let start = clock.now
        _ = try await submit(.lookupAccounts, payload: probe, timeout: timeout)
        return clock.now - start
    }

    public func lookupAccounts(_ ids: [UInt128]) async throws(TBError) -> [Account] {
        guard !ids.isEmpty else { return [] }
        return Wire.accounts(try await submit(.lookupAccounts, payload: Wire.ids(ids)))
    }

    public func lookupTransfers(_ ids: [UInt128]) async throws(TBError) -> [Transfer] {
        guard !ids.isEmpty else { return [] }
        return Wire.transfers(try await submit(.lookupTransfers, payload: Wire.ids(ids)))
    }

    public func lookupAccount(_ id: UInt128) async throws(TBError) -> Account {
        guard let a = try await lookupAccounts([id]).first else { throw .notFound("Account \(id)") }
        return a
    }

    public func lookupTransfer(_ id: UInt128) async throws(TBError) -> Transfer {
        guard let t = try await lookupTransfers([id]).first else { throw .notFound("Transfer \(id)") }
        return t
    }

    /// Account and transfer id spaces are independent; an account wins if both exist.
    public func lookupID(_ id: UInt128) async throws(TBError) -> LookupResult {
        if let a = try await lookupAccounts([id]).first { return .account(a) }
        if let t = try await lookupTransfers([id]).first { return .transfer(t) }
        return .none
    }

    public func queryAccounts(_ filter: QueryFilter) async throws(TBError) -> [Account] {
        Wire.accounts(try await submit(.queryAccounts, payload: Wire.encode(filter)))
    }

    public func queryTransfers(_ filter: QueryFilter) async throws(TBError) -> [Transfer] {
        Wire.transfers(try await submit(.queryTransfers, payload: Wire.encode(filter)))
    }

    public func accountTransfers(_ accountID: UInt128, _ filter: AccountFilter) async throws(TBError) -> [Transfer] {
        Wire.transfers(try await submit(.getAccountTransfers, payload: Wire.encode(filter, accountID: accountID)))
    }

    public func accountBalances(_ accountID: UInt128, _ filter: AccountFilter) async throws(TBError) -> [Balance] {
        Wire.balances(try await submit(.getAccountBalances, payload: Wire.encode(filter, accountID: accountID)))
    }

    /// Finds a transfer by id and reports whether it involves `accountID`.
    /// `get_account_transfers` cannot filter by transfer id, but a lookup is exact and O(1).
    public func findTransfer(_ id: UInt128, onAccount accountID: UInt128) async throws(TBError) -> AccountTransferMatch {
        guard let t = try await lookupTransfers([id]).first else { return .notFound }
        return t.debitAccountID == accountID || t.creditAccountID == accountID ? .onAccount(t) : .otherAccounts(t)
    }

    /// The transfer that has exactly this timestamp, e.g. the source of a balance row.
    public func transfer(atTimestamp ts: UInt64) async throws(TBError) -> Transfer? {
        try await queryTransfers(QueryFilter(timestampMin: ts, timestampMax: ts, limit: 1)).first
    }
}

File: cli/tbclient/src/queries.rs
//! The six read operations, and the two ways of asking about one account.

use crate::client::{Client, DEFAULT_TIMEOUT, TbError};
use crate::ffi::Operation;
use crate::models::{Account, AccountFilter, Balance, QueryFilter, Transfer};
use crate::wire;

/// What an id turned out to name.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Found {
    Account(Account),
    Transfer(Transfer),
    Nothing,
}

impl Client {
    pub fn lookup_accounts(&self, ids: &[u128]) -> Result<Vec<Account>, TbError> {
        if ids.is_empty() {
            return Ok(Vec::new());
        }
        let reply = self.submit(Operation::LookupAccounts, wire::ids(ids), DEFAULT_TIMEOUT)?;
        wire::accounts(&reply)
    }

    pub fn lookup_transfers(&self, ids: &[u128]) -> Result<Vec<Transfer>, TbError> {
        if ids.is_empty() {
            return Ok(Vec::new());
        }
        let reply = self.submit(Operation::LookupTransfers, wire::ids(ids), DEFAULT_TIMEOUT)?;
        wire::transfers(&reply)
    }

    pub fn account_transfers(&self, filter: &AccountFilter) -> Result<Vec<Transfer>, TbError> {
        let reply = self.submit(
            Operation::GetAccountTransfers,
            wire::account_filter(filter),
            DEFAULT_TIMEOUT,
        )?;
        wire::transfers(&reply)
    }

    /// Only accounts created with the `history` flag have balances to return; for the rest the
    /// cluster answers with nothing, which is not an error.
    pub fn account_balances(&self, filter: &AccountFilter) -> Result<Vec<Balance>, TbError> {
        let reply = self.submit(
            Operation::GetAccountBalances,
            wire::account_filter(filter),
            DEFAULT_TIMEOUT,
        )?;
        wire::balances(&reply)
    }

    pub fn query_accounts(&self, filter: &QueryFilter) -> Result<Vec<Account>, TbError> {
        let reply = self.submit(
            Operation::QueryAccounts,
            wire::query_filter(filter),
            DEFAULT_TIMEOUT,
        )?;
        wire::accounts(&reply)
    }

    pub fn query_transfers(&self, filter: &QueryFilter) -> Result<Vec<Transfer>, TbError> {
        let reply = self.submit(
            Operation::QueryTransfers,
            wire::query_filter(filter),
            DEFAULT_TIMEOUT,
        )?;
        wire::transfers(&reply)
    }

    /// Resolves an id the user typed, which could name either kind. Accounts are asked first
    /// because that is the commoner case, and an id is never both.
    pub fn lookup_id(&self, id: u128) -> Result<Found, TbError> {
        if let Some(account) = self.lookup_accounts(&[id])?.first() {
            return Ok(Found::Account(*account));
        }
        if let Some(transfer) = self.lookup_transfers(&[id])?.first() {
            return Ok(Found::Transfer(*transfer));
        }
        Ok(Found::Nothing)
    }
}
